Transform loops so that they have a single exit out of the loop header

bgogul opened the pull request:
<!-- What's in this pull request? -->
Fix for [SR-7765](https://bugs.swift.org/browse/SR-7765). This is a first cut of the necessary transformations to convert loops into a form that can be translated directly to XF-TLA while loops. 

This does not work with labeled breaks yet. I will fix them in a subsequent PR. 
<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-7765](https://bugs.swift.org/browse/SR-7765).
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

bgogul committed:
SESE fix using Undefs

bgogul committed:
Adds a command line flag to guard single exit.

bgogul committed:
Fix a bug with nested loops.

bgogul committed:
Fix a bug with exitIndices being not contiguous

bgogul committed:
Added a test for SESE loop canonicalization.

bgogul commented:
@swift-ci please test tensorflow

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+    assert(headerExit == nullptr && "Loop header has more than one exit node.");
+    headerExit = succ;
+  }
+  if (headerExit != nullptr) {
```
shouldn't we assert headerExit is not NULL here?

bgogul replied:
No, because the header may not be an exiting block. canonicalizeLoops does not ensure that.

mhong replied:
you are right -- i guess an example is:

```swift
while true {
   ...
   if ... {
     break
  }
}

can we add a comment?

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  }
+  unsigned nextExitIndex = 1;
+  for (const auto &edge : edgesToFix) {
+    SILBasicBlock *src = const_cast<SILBasicBlock *>(edge.first);
```
is it possible to not use const in edge in the first place?

bgogul replied:
Unfortunately,  getExitEdges returns a set of const SILBasicBlock* pairs. The only option is to reimplement getExitEdges here, which seemed worse than a const cast. :(

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  for (const auto &edge : edgesToFix) {
+    SILBasicBlock *src = const_cast<SILBasicBlock *>(edge.first);
+    SILBasicBlock *tgt = const_cast<SILBasicBlock *>(edge.second);
+    SILBasicBlock *newTgt = loop->contains(src) ? latchBlock : newHeader;
```
this stmt is not obvious -- consider adding a comment. more generally provide context on what `edgesToFix` mean and why we use that.

bgogul replied:
Added comments describing the meaning of edgesToFix. I also changed loop->contains(src) to  (src == preheader), which is easier to read and understand.

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+      if (DI.properlyDominates(escapingValue, src->getTerminator())) {
+        newArgs.push_back(escapingValue);
+      } else if (src != preheader) {
+        // if src is not the preheader than new header arguments are available.
```
than -> then

bgogul replied:
Done.

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+        // if src is not the preheader than new header arguments are available.
+        newArgs.push_back(newHeader->getArgument(argIndex));
+      } else {
+        newArgs.push_back(getUndef(escapingValue->getType()));
```
i don't fully understand why and how we use undef -- some more context would be nice.

bgogul replied:
Added a brief explanation here.

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+        {exitIndexArg,
+         createTFIntegerConst(builder, headerLocation, /*bitwidth*/ 32,
+                              exitIndices.lookup(trueBlock)),
+         builder.createStringLiteral(headerLocation, StringRef(ALL_DEVICES),
```
can we use handleDevicePlacement in TFUtilities instead?

mhong reviewed (commented):
I don't fully understand this patch yet. Clearly it has a lot of complexity as the problem is inherently non-trivial! Some more context (a 1-pager doc or large comment blocks) would be very helpful with understanding the design.

Sorry I left a large # of comments -- from high level design questions to nitty-gritty coding details. Please feel free to defer addressing the latter if you prefer, and help me understand the design more first. Thanks.

lattner commented:
@bgogul, let me know when you and hongm converge and I'll take a look.

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+      // Increment index as we inserted a new entry into the table.
+      ++nextExitIndex;
+    }
+    auto kvPair = *(emplaceResult.first);
```
Redundant parentheses?

bgogul replied:
Done

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+    SILBasicBlock *newBlock = createBlockOutsideLoop();
+    SILBasicBlock *trueBlock = *curBlockIter++;
+    builder.setInsertionPoint(newBlock);
+    SILValue condTensor = builder.createBuiltin(
```
Would it make sense to use `graph_op` instruction instead? We are moving towards `graph_op` for all code paths after deabstraction.

bgogul replied:
Done.

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  SILType elementType = element->getType();
+  // Literals take attributes specifying the dtype, value, and device.
+  std::string opName("__tfop_Const,dtype$dtype,value$tensor,__device");
+  SmallVector<SILValue, 8> operands;
```
There are exactly three operands, so there's no need for `SmallVector`. A C array `SILValue operands[]` would be simpler and more efficient.

bgogul replied:
Done.

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+/// Appends the given arguments to the given edge. Deletes the old TermInst
+/// and returns a new TermInst with the appropriate number of args.
+TermInst* appendArguments(TermInst *termInst, SILBasicBlock *target,
+                     ArrayRef<SILValue> newArgs) {
```
Align with the first argument.

bgogul replied:
Done.

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+
+/// Appends the given arguments to the given edge. Deletes the old TermInst
+/// and returns a new TermInst with the appropriate number of args.
+TermInst* appendArguments(TermInst *termInst, SILBasicBlock *target,
```
The local convention is `TermInst *appendArguments`

bgogul replied:
Done.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -261,8 +634,8 @@ void SESERegionBuilder::processLoop(SILLoop *loop) {
 
-  // Loop canonicalization also gives us the property that exits out of the loop
-  // have critical edges split, and that any exit block jumps to a block
+  // Loop canonicalization also gives us the property that exits out of the
```
I simply did clang-format without any additional configuration. However,  it seems to use the llvm style automatically.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -19,14 +19,26 @@
 using namespace tf;
 
+static llvm::cl::opt<bool> TFEnsureSingleLoopExit(
+    "tf-ensure-single-loop-exit", llvm::cl::init(false),
```
Yes, this flag is just for migration purposes. Will remove it once I land the PR that fixes undefs while lowering the graph.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+/// to the appropriate exit block. Updates the Dominanceinfo, PostDominanceinfo,
+/// and LoopInfo as well.
+bool SESERegionBuilder::ensureSingleExitFromLoops() {
+  // TODO: update changed appropriately.
```
Oops, this is a stale comment.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  llvm::SmallVector<std::pair<SILLoop *, bool>, 16> workList;
+  for (auto *L : LI.getTopLevelLoops())
+    workList.push_back({L, L->empty()});
+  while (workList.size() != 0) {
```
Sure. Done.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+bool SESERegionBuilder::ensureSingleExitFromLoop(SILLoop* loop) {
+  // Return if the loop is already in the required form.
+  if (loop->getExitBlock() && loop->getExitingBlock() &&
+      loop->getExitingBlock() == loop->getHeader()) {
```
No, even if the loop has a single exit and exiting block, the exiting block may not be the header. (canonicalizeLoop does not ensure that property.)

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  // TODO: update changed appropriately.
+  // Visit the loop nest hierarchy bottom up.
+  bool changed = false;
+  llvm::SmallVector<std::pair<SILLoop *, bool>, 16> workList;
```
I don't think having two variables will help that much. I just followed the pattern in canonicalizeAllLoops

bgogul commented on test/TensorFlow/sese_loop_canonicalization.sil:
```diff
@@ -0,0 +1,219 @@
+
+//-- New Header simply checks flag
+// CHECK: bb9([[HDR_ORIG_ARG]] : $Builtin.Int32, {{.*}} : $Builtin.Int32, [[HDR_ESCAPING_ARG]] : $Builtin.Int32, [[HDR_EXIT_ARG:%.*]] : $TensorHandle<Builtin.Int32>, [[HDR_FLAG_ARG:%.*]] : $TensorHandle<Builtin.Int1>):
+// CHECK:  [[B:%.*]] = builtin "tf_tensor_to_i1"([[HDR_FLAG_ARG]] : $TensorHandle<Builtin.Int1>) : $Builtin.Int1
```
I am treating A's and B's like local variables. :) Lit seems to rebind the names without any issues.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+      llvm::for_each(inst.getResults(), save_escaping);
+    }
+  }
+  llvm::outs() << "The following are live outside the loop:\n";
```
Removed this completely.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+                              exitIndices.lookup(trueBlock)),
+         builder.createStringLiteral(headerLocation, StringRef(ALL_DEVICES),
+                                     StringLiteralInst::Encoding::UTF8)});
+    SILValue condValue = builder.createBuiltin(
```
Done.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  // As a simplification, simply 0 as exit index for edges to the header
+  // even though header is not an exit block.
+  exitIndices.insert({header, 0});
+  // Identify the exit from the header and also set it to zero index.
```
I have changed the comments a bit. PTAL.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  }
+
+  llvm::DenseMap<SILBasicBlock *, intmax_t> exitIndices;
+  // As a simplification, simply 0 as exit index for edges to the header
```
I have changed the comments a bit. PTAL.

bgogul committed:
test changes

bgogul committed:
Addressed most of the readability comments.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  loop->addBasicBlockToLoop(newHeader, LI.getBase()); // Should be done first.
+  // Clone arguments and change all uses to the new header's arguments.
+  newHeader->cloneArgumentList(header);
+  SmallVector<SILValue, 8> headerArgs;
```
Tell me about it. :( There are some adapters like [ArrayRefView](https://github.com/apple/swift/blob/d6dad24ffee15ffbfbabc37dbbc573514a70884d/include/swift/Basic/ArrayRefView.h), which might help here. However, I haven't spent the time understanding the API yet.

bgogul commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+}
+
+
+bool SESERegionBuilder::ensureSingleExitFromLoop(SILLoop* loop) {
```
This function should be split up into multiple functions. That should improve the readability of the code.  Let me take a stab at it.

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +264,364 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  ASTContext& context = builder.getASTContext();
+  SILType elementType = element->getType();
+  // Literals take attributes specifying the dtype, value, and device.
+  std::string opName("__tfop_Const,dtype$dtype,value$tensor,__device");
```
The `graph_op` name should be `Const`, not the "__tfop_" encoding that builtins are using. Dtypes are already implied when you specify operands. `__device` is an attribute in this instruction.

rxwei replied:
See usage here: https://github.com/apple/swift/blob/39ced43dd4c59095fe0b7135ccbf42098fd7b410/test/TensorFlow/integration.swift#L64

bgogul replied:
Thanks for the pointer. I think I am not setting the attributes properly as well. Is there a code snippet in C++ where "Const" graph operation is created?

rxwei replied:
https://github.com/apple/swift/blob/39ced43dd4c59095fe0b7135ccbf42098fd7b410/lib/SILOptimizer/Mandatory/TFPartition.cpp#L2977

bgogul replied:
Thanks for the pointer. I changed all the relevant builtin instructions to graph operations. PTAL.

bgogul committed:
Changed builtins to graph operations.

bgogul committed:
Refactor the loooong transformation function into a bunch of smaller functions.

bgogul commented:
@mhong @rxwei, I updated the PR with the following changes:
  - use graph operations instead of builtins
  - refactor the transformation code into a bunch of little functions, which hopefully makes it easy to read. 

PTAL.

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+      SILType::getBuiltinIntegerType(bitwidth, builder.getASTContext());
+  // Literals take attributes specifying the dtype, value, and device.
+  std::string opName("Const");
+  SmallVector<GraphOperationAttribute, 3> attributes;
```
This can be a C array also.

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+                        SymbolicValue::getInteger(APInt(bitwidth, value),
+                                                  context.getAllocator())});
+  graphConfiguration->handleDevicePlacement(
+      "Const",
```
Nice!

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+
+  size_t oldHeaderNumArgs =
+      newHeader->getNumArguments() -
+      (escapingValues.size() + /* exitIndex, stayLoop*/ 2);
```
What does "stayLoop" mean?

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+
+  llvm::DenseMap<SILBasicBlock *, intmax_t> exitIndices;
+
+  size_t oldHeaderNumArgs =
```
It's more common to use `unsigned` around here in the swift codebase.

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+
+  // Identify the exit from the header (if any) and assign '0' as its index.
+  SILBasicBlock* headerExit = nullptr;
+  for (SILBasicBlock* succ : header->getSuccessorBlocks()) {
```
Minor nit: `SILBasicBlock *`

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+
+  llvm::DenseMap<SILBasicBlock *, intmax_t> exitIndices;
+
+  size_t oldHeaderNumArgs =
```
should we save oldHeaderNumArgs at initialize() time, and just do an assert for sanity check here?

mhong replied:
it just feels round-about that we are recovering / reverse-engineering some info that we could get from the input. instead we are mutating the input state, and then cancelling out its effect to get that same info.

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+    builder.setInsertionPoint(newBlock);
+
+    // Create a condition to compare exitIndex to a constant
+    std::string equalOpName("Equal");
```
this chunk of code looks similar to code in `handleSendRecvForTerminator()` in TFPartition.cpp. should we do some refactoring?

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+/// Build a demuxing if..then..else block that can be used as a single
+/// exit block. This will not have any effect if there is a single exit
+/// block already.
+SILBasicBlock *SingleExitLoopTransformer::createNewExitBlock(
```
we are creating more than a single exit block -- maybe call this function sth like `createNewExitBlockWithDemux()`?

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+        condBranch->getTrueBBCount(), condBranch->getFalseBBCount());
+  } else {
+    // At the moment we can only add arguments to br and cond_br.
+    llvm_unreachable("Can't add argument to terminator");
```
can we do assert on CondBr on the previous `else if` case, and remove this `else` case?

mhong replied:
It makes it clear that we only have two possible cases. With this extra else body, it's not clear if we are missing some impl code, to be added in the future. Also the code gets longer.

bgogul replied:
Fair enough. Done. :)

mhong reviewed (commented):
Overall I'm able to follow this PR. The doc and comment blocks help a lot. Thank you! The main open design question is whether we can remove undefs -- we can discuss outside this PR.

I believe we'll just need one more round of review here, including the tests.

bgogul committed:
Addressed another round of review comments:
- Moving around some code
  - Renaming some functions and variables
  - Adding some comments
  - Added tests to make sure exit index and flag are set correctly.

bgogul reviewed (commented):
I have addressed all the comments. PTAL. 

One thing that is remaining is adding end-to-end tests. However, that will have to wait until we get rid of undefs.

bgogul committed:
Fixed a minor typo

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +261,470 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+/// For loops in the SIL Function with multiple exit block successors,
+/// transforms the loop by introducing a common exit switch block that demuxes
+/// to the appropriate exit block. Updates the Dominanceinfo, PostDominanceinfo,
+/// and LoopInfo as well.
```
Right. I had a typo: now -> not.

the comment text is out-dated.

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
 
+/// Create a TF handle of the appropriate integer type for the given bitwith
+/// that is initialized to the given value.
+SILValue createTFIntegerConst(GraphGlobalConfiguration& graphConfiguration,
```
add static

lattner replied:
+1

mhong commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+}
+
+SILBasicBlock *
+SingleExitLoopTransformer::createLatchBlock(SILBasicBlock *newHeader) {
```
nit: the previous func is called `createNewHeader`. should this be `createNewLatch`?

mhong reviewed (approved):
Thank you, and great work for tackling one of the most intricate problems in our project!

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  /// the new block and the phi argument corresponding to the exitIndex.
+  std::pair<SILBasicBlock *, SILValue> createNewHeader();
+
+  /// Create a new latch block and clones all the arguments from new header.
```
"clones" -> "clone", since "create" was used instead of "creates"

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+
+  void initialize();
+
+  /// Create a new header for the loop and returns a pair consisting of
```
"returns" -> "return"

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+
+  // Configuration for graph construction.
+  GraphGlobalConfiguration *graphConfiguration;
+  DominanceInfo* DI;
```
`DominanceInfo *` for consistency with local code

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  // Configuration for graph construction.
+  GraphGlobalConfiguration *graphConfiguration;
+  DominanceInfo* DI;
+  SILLoopInfo* LI;
```
`SILLoopInfo *`

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+void SingleExitLoopTransformer::initialize() {
+  // Remember some information from the loop before it gets transformed.
+
+  // exit Blocks
```
"Exit blocks"?

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+/// and returns a new TermInst with the appropriate number of args.
+TermInst *SingleExitLoopTransformer::appendArguments(
+    TermInst *termInst, SILBasicBlock *target, ArrayRef<SILValue> newArgs) {
+  SILBuilder builder(termInst);
```
Since the entire transformation is local to a function, it would be better to define a single builder in the Transformer class.

bgogul replied:
Ah, good idea. Will do in a subsequent PR as a cleanup.

rxwei commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  // Create a new latch block.
+  SILBasicBlock *latchBlock = currentFn->createBasicBlock();
+  latchBlock->cloneArgumentList(newHeader);
+  SILBuilder  builder(latchBlock);
```
Extra space.

rxwei reviewed (approved):
LGTM once you and @mhong converge on the algorithm. Most of my comments are style-related nits.

lattner commented on include/swift/SIL/SILFunction.h:
```diff
@@ -927,6 +927,9 @@ class SILFunction
   /// block inside.  This depends on there being a 'dot' and 'gv' program in
   /// your path.
   void viewCFG() const;
+  // Like the ViewCFG, but the graph does not show the basic block contents.
```
Ok, but should be marked as SWIFT_ENABLE_TENSORFLOW and upstreamed asap :-)

bgogul replied:
Oops, this is already up streamed. I forgot to remove this patch before sending out the PR.

lattner commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -19,14 +19,27 @@
 
 #include "TFCanonicalizeCFG.h"
+#include "TFUtilities.h"
+#include "llvm/ADT/iterator_range.h"
```
Please move the llvm include below the swift includes since LLVM is a lower layer than swift.

lattner commented on lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp:
```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+                        SymbolicValue::getInteger(APInt(bitwidth, value),
+                                                  context.getAllocator())});
+  graphConfiguration.handleDevicePlacement(
+      "Const",
```
Pass opName here, or eliminate opName?

bgogul commented on lib/SIL/SILFunction.cpp:
```diff
@@ -434,17 +434,28 @@ TargetFunction("view-cfg-only-for-function", llvm::cl::init(""),
-
-void SILFunction::viewCFG() const {
+namespace {
+void viewCFGHelper(const SILFunction* F, bool skipBBContents) {
```
This should not be here. I removed it in the latest patch.

bgogul commented on test/TensorFlow/sese_loop_canonicalization.sil:
```diff
@@ -0,0 +1,219 @@
+import TensorFlow
+
+/*
+public func loopTest(breakCount:Int32) -> Int32 {
```
Will do this once we get rid of undefs.

bgogul committed:
Final round of changes

bgogul reviewed (commented):
I have addressed all the comments and will merge the PR shortly.

rxwei commented:
Congrats on landing this :)
